src/client/InteractionClient.js

```diff
@@ -0,0 +1,211 @@
+            default: o.default,
+            required: o.required,
+            choices: o.choices,
+            options: o.options?.map(m),
+          };
+        }),
+      },
+    });
+    return new ApplicationCommand(this, c, guildID);
+  }
+
+  handle(data) {
+    switch (data.type) {
+      case InteractionType.PING:
+        return {
+          type: InteractionResponseType.PONG,
+        };
+      case InteractionType.APPLICATION_COMMAND: {
+        let timedOut = false;
+        let resolve;
+        const directPromise = new Promise(r => {
+          resolve = r;
+          this.client.setTimeout(() => {
+            timedOut = true;
+            r({
+              type: InteractionResponseType.ACKNOWLEDGE_WITH_SOURCE,
+            });
+          }, 250);
+        });
+
+        const syncHandle = {
+          acknowledge({ hideSource }) {
+            if (!timedOut) {
+              resolve({
+                type: hideSource
+                  ? InteractionResponseType.ACKNOWLEDGE
+                  : InteractionResponseType.ACKNOWLEDGE_WITH_SOURCE,
+              });
+            }
+          },
+          reply(resolved) {
+            if (timedOut) {
+              return false;
+            }
+            resolve({
+              type: resolved.hideSource
+                ? InteractionResponseType.CHANNEL_MESSAGE
+                : InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
+              data: resolved.data,
+            });
+            return true;
+          },
+        };
+
+        const interaction = new CommandInteraction(this.client, data, syncHandle);
+
+        /**
+         * Emitted when an interaction is created.
+         * @event Client#interactionCreate
+         * @param {Interaction} interaction The interaction which was created.
+         */
+        this.client.emit(Events.INTERACTION_CREATE, interaction);
+
+        return directPromise;
+      }
+      default:
+        throw new RangeError('Invalid interaction data');
+    }
+  }
+
+  /**
+   * An express-like middleware factory which can be used
+   * with webhook interactions.
+   * @returns {Function} The middleware function.
+   */
+  middleware() {
+    return async (req, res) => {
+      const timestamp = req.get('x-signature-timestamp');
+      const signature = req.get('x-signature-ed25519');
+
+      const chunks = [];
+      for await (const chunk of req) {
+        chunks.push(chunk);
+      }
+      const body = Buffer.concat(chunks);
+
+      if (sodium === undefined) {
+        sodium = require('../util/Sodium');
+      }
+      if (
+        !sodium.methods.verify(
+          Buffer.from(signature, 'hex'),
+          Buffer.concat([Buffer.from(timestamp), body]),
+          this.publicKey,
+        )
+      ) {
+        res.status(401).end();
+        return;
+      }
+
+      const data = JSON.parse(body.toString());
+
+      const result = await this.handle(data);
+      res.status(200).end(JSON.stringify(result));
+    };
+  }
```

Comment:
This **shouldn't** be in discord.js at all, it's not up to discord.js to provide this middleware. 

Moreover, this middleware is only compatible with ExpressJS. ExpressJS is one of the worst HTTP "micro-framework". So well, just because most noobs use ExpressJS, doesn't mean you should provide middleware, and even worse, middleware that is only compatible with ExpressJS, a terrible micro-framework.

In the worst case, you can put this middleware in an example, or the documentation.
But in any case, it's not to discord.js to provide middleware and even less for ExpressJS.

Reply:
This change has every right to be in discord.js. So what if people use Express.js? As long as it works and as long as it works well. It's not discord.js' **job** to provide the middleware, but the community chose to add it.

Reply:
First, **ExpressJS is low-scope**, that is to say, that it offers nothing more than a router.
You rarely need a simple router to create your site or application. ExpressJS provides you with very few features to create a complete and good quality site or application. If you only need a simple router, you can look at alternatives such as Fastify.

Secondly, **ExpressJS has not been updated for 3 years**, it doesn't know how to handle asynchronous and **will easily crash** your applications if you are not careful.

Thirdly, even though ExpressJS is presented as a fast framework, it is the **slowest** framework that exists in the Node.js universe.

For a good and complete alternative, you can look at the framework **[AdonisJS](https://preview.adonisjs.com)**.

Reply:
> This change has every right to be in discord.js. So what if people use Express.js? As long as it works and as long as it works well. It's not discord.js' job to provide the middleware, but the community chose to add it.

**This is not to discord.js to make a middleware and even less for noobs and a __bad framework__**

Reply:
I support @Mesteery.
If all libraries used Express syntax and structure then it would make sense, but not in this case.

Reply:
> 
> 
> This **shouldn't** be in discord.js at all, it's not up to discord.js to provide this middleware.
> 
> Moreover, this middleware is only compatible with ExpressJS. ExpressJS is one of the worst HTTP "micro-framework". So well, just because most noobs use ExpressJS, doesn't mean you should provide middleware, and even worse, middleware that is only compatible with ExpressJS, a terrible micro-framework.
> 
> In the worst case, you can put this middleware in an example, or the documentation.
> But in any case, it's not to discord.js to provide middleware and even less for ExpressJS.

You talk about noobs while commenting on a PR of discord.js. Calm down 🤙

Reply:
What is your problem ?

Reply:
> 
> 
> What is your problem ?

The question is what is yours, you try and take a shit on Express.js every chance you get, I understand you may dislike or have experience with Express.js not being optimal, but you expressed that so rudely, so I wonder, what is your problem?

Reply:
> 
> 
> This **shouldn't** be in discord.js at all, it's not up to discord.js to provide this middleware.
> 
> Moreover, this middleware is only compatible with ExpressJS. ExpressJS is one of the worst HTTP "micro-framework". So well, just because most noobs use ExpressJS, doesn't mean you should provide middleware, and even worse, middleware that is only compatible with ExpressJS, a terrible micro-framework.
> 
> In the worst case, you can put this middleware in an example, or the documentation.
> But in any case, it's not to discord.js to provide middleware and even less for ExpressJS.

Behind of all this btw, this could go either ways, discord.js can provide support for both methods with webservers and using the gateway, but yes I feel like throwing a middleware into the mix of a draft PR is not gonna be ideal. So yeah, I agree with you.

Reply:
Yes, the 'noobs' may have been a bit aggressive. But I don't share my dissatisfaction with ExpressJS, it's a general opinion because ExpressJS is really bad, and I assure you that I'm not the only one to advise against it.

And again, other than the fact that it's middleware for ExpressJS, it wouldn't have anything to do with discord.js. It's not up to discord.js to provide this "tool".

This middleware can also be completely useless for most people, for several reasons: most bots don't have a web application or don't use ExpressJS.
Therefore, it would be completely inappropriate to provide middleware that is only usable with ExpressJS and that few people will use it.

Reply:
> 
> 
> Yes, the 'noobs' may have been a bit aggressive. But I don't share my dissatisfaction with ExpressJS, it's a general opinion because ExpressJS is really bad, and I assure you that I'm not the only one to advise against it.
> 
> And again, other than the fact that it's middleware for ExpressJS, it wouldn't have anything to do with discord.js. It's not up to discord.js to provide this "tool".
> 
> This middleware can also be completely useless for most people, for several reasons: most bots don't have a web application or don't use ExpressJS.
> Therefore, it would be completely inappropriate to provide middleware that is only usable with ExpressJS and that few people will use it.

Makes sense, web server stuff should be left out of this draft pr for everyone's sanity